xchange-stream-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingAdapters.java

```diff
@@ -19,58 +15,96 @@
+        adaptLimitOrders(instrument, orderType, currentNode.get(key)).forEachRemaining(limitOrder -> {
+            target.removeIf(it -> it.getLimitPrice().compareTo(limitOrder.getLimitPrice()) == 0);
+            if (limitOrder.getOriginalAmount().compareTo(BigDecimal.ZERO) != 0) {
+                target.add(limitOrder);
+            }
+        });
+        while (target.size() > depth) {
+            LimitOrder last = target.last();
+            target.remove(last);
+        }
+    }
 
-  public static OrderBook adaptOrderbookMessage(
-      OrderBook orderBook, Instrument instrument, ArrayNode arrayNode) {
-    Streams.stream(arrayNode.elements())
-        .filter(JsonNode::isObject)
-        .forEach(
-            currentNode -> {
-              for (String key : BID_KEYS) {
-                if (currentNode.has(key)) {
-                  adaptLimitOrders(instrument, Order.OrderType.BID, currentNode.get(key))
-                      .forEach(orderBook::update);
+    public static OrderBook adaptOrderbookMessage(int depth, TreeSet<LimitOrder> bids, TreeSet<LimitOrder> asks, Instrument instrument, ArrayNode arrayNode) {
+        final AtomicLong expectedChecksum = new AtomicLong(0);
+        final AtomicReference<Date> lastTime = new AtomicReference<>(Date.from(Instant.EPOCH));
+        final boolean awaitingSnapshot = (bids.isEmpty() && asks.isEmpty());
+        arrayNode.elements().forEachRemaining(currentNode -> {
+            if (awaitingSnapshot) {
+                if (currentNode.has(BID_SNAPSHOT) && currentNode.has(ASK_SNAPSHOT)) {
+                    LOG.info("Received {} snapshot, clearing book", instrument);
+                    updateInBook(depth, instrument, Order.OrderType.BID, currentNode, BID_SNAPSHOT, bids);
+                    updateInBook(depth, instrument, Order.OrderType.ASK, currentNode, ASK_SNAPSHOT, asks);
                 }
-              }
-              for (String key : ASK_KEYS) {
-                if (currentNode.has(key)) {
-                  adaptLimitOrders(instrument, Order.OrderType.ASK, currentNode.get(key))
-                      .forEach(orderBook::update);
+            } else {
+                if (currentNode.has(BID_UPDATE)) {
+                    updateInBook(depth, instrument, Order.OrderType.BID, currentNode, BID_UPDATE, bids);
                 }
-              }
-            });
-    return new OrderBook(
-        orderBook.getTimeStamp(),
-        Lists.newArrayList(orderBook.getAsks()),
-        Lists.newArrayList(orderBook.getBids()),
-        true);
-  }
+                if (currentNode.has(ASK_UPDATE)) {
+                    updateInBook(depth, instrument, Order.OrderType.ASK, currentNode, ASK_UPDATE, asks);
+                }
+            }
+            if (!awaitingSnapshot && currentNode.has(CHECKSUM)) {
+                expectedChecksum.set(currentNode.get(CHECKSUM).asLong());
+            }
+        });
+        if ( awaitingSnapshot && bids.isEmpty() && asks.isEmpty()){
```

Comment:
remove spacing on left

Reply:
also why do you need the additional awaiting snapshot boolean check

Reply:
We probably don't, i'll remove that, low impact as it's logging only anyway.